sscanss/core/geometry/intersection.py

```diff
@@ -374,3 +381,74 @@ def volume_plane_intersection(volume, plane, resolution=1024):
 
     rect = (*min_limits[:2], *(max_limits[:2] - min_limits[:2]))
     return VolumeSlice(image_slice.transpose().copy(), rect)
+
+
+def volume_ray_intersection(start, end, volume):
+    """Finds point where a ray intersects a volume - uses same algorithm as in the fragment shader for volumes
+    :param volume: the volume to find intersection with
+    :type volume: Volume
+    :param start: start of the ray
+    :type start: Vector3
+    :param end: end of the ray
+    :type end: Vector3
+    :return: the point where ray intersects the volume
+    :rtype: Vector3
+    """
+
+    # First get the mesh of bounding box of the volume
+    width = volume.voxel_size[0] * volume.data.shape[0]
+    height = volume.voxel_size[1] * volume.data.shape[1]
+    depth = volume.voxel_size[2] * volume.data.shape[2]
+
+    bounding_box_mesh = create_cuboid(width, depth, height)
+    bounding_box_mesh.transform(volume.transform_matrix)
+
+    # Find where does the ray intersect the volume's planes (should be 2 points)
+    line_segment_direction = end - start
+    line_segment_length = line_segment_direction.length
+    line_segment_direction /= line_segment_length
+    faces = bounding_box_mesh.vertices[bounding_box_mesh.indices].reshape(-1, 9)
+
+    intersection_distances = segment_triangle_intersection(start, line_segment_direction, line_segment_length, faces)
+
+    if len(intersection_distances) == 0:
+        return None
+    if len(intersection_distances) == 1:
+        intersection_distances.append(0)
+
+    intersection_distances.sort()
+    points = []
+    translate = Vector3([width / 2, height / 2, depth / 2])
+    for distance in intersection_distances:
+        point = start + line_segment_direction * distance
+        point = Vector3((np.linalg.inv(volume.transform_matrix) @ Vector4([point.x, point.y, point.z, 0]))[:3])
+        point = point + translate
+        points.append(point)
+
+    close_point = points[0]
+    far_point = points[1]
+
+    inside_segment = far_point - close_point
+    num_of_steps = 10000
+    step_vector = inside_segment / num_of_steps
+
+    x = np.linspace(0, (volume.data.shape[0]) * volume.voxel_size[0], num=volume.data.shape[0])
+    y = np.linspace(0, (volume.data.shape[1]) * volume.voxel_size[1], num=volume.data.shape[1])
+    z = np.linspace(0, (volume.data.shape[2]) * volume.voxel_size[2], num=volume.data.shape[2])
+    interpolator = RegularGridInterpolator((x, y, z), volume.data, bounds_error=False, fill_value=0.0)
```

Comment:
please investigate `map_coordinates` might be faster than `RegularGridInterpolator`